retrieval/targetmanager.go

```diff
@@ -21,132 +23,346 @@ import (
-				m.AddTarget(job, target)
+// targetsFromGroup builds targets based on the given TargetGroup and config.
+func (tm *TargetManager) targetsFromGroup(tg *config.TargetGroup, cfg config.JobConfig) ([]Target, error) {
+	targets := make([]Target, 0, len(tg.Targets))
+	for i, labels := range tg.Targets {
+		for ln, lv := range tg.Labels {
+			if _, ok := labels[ln]; !ok {
+				labels[ln] = lv
 			}
 		}
+		for ln, lv := range tm.globalLabels {
+			if _, ok := labels[ln]; !ok {
+				labels[ln] = lv
+			}
+		}
+		address, ok := labels[clientmodel.AddressLabel]
+		if !ok {
+			return nil, fmt.Errorf("Instance %d in target group %s has no address", i, tg)
+		}
+		// Remove hidden labels from final target.
+		for ln := range labels {
+			if len(strings.TrimLeft(string(ln), "_")) == len(ln)-1 {
+				delete(labels, ln)
+			}
+		}
+		targets = append(targets, NewTarget(string(address), cfg, labels))
 	}
+	return targets, nil
 }
 
-func (m *targetManager) Stop() {
-	m.Lock()
-	defer m.Unlock()
+// ProvidersFromConfig returns all TargetProviders configured in cfg.
+func ProvidersFromConfig(cfg config.JobConfig) ([]TargetProvider, error) {
+	var providers []TargetProvider
 
-	glog.Info("Stopping target manager...")
-	var wg sync.WaitGroup
-	for j, p := range m.poolsByJob {
-		wg.Add(1)
-		go func(j string, p *TargetPool) {
-			defer wg.Done()
-			glog.Infof("Stopping target pool %q...", j)
-			p.Stop()
-			glog.Infof("Target pool %q stopped.", j)
-		}(j, p)
+	if name := cfg.GetSdName(); name != "" {
+		dnsSD := discovery.NewDNSDiscovery(name, cfg.SDRefreshInterval())
+		providers = append(providers, dnsSD)
 	}
-	wg.Wait()
-	glog.Info("Target manager stopped.")
+
+	if tgs := cfg.TargetGroups(); tgs != nil {
+		static := NewStaticProvider(tgs)
+		providers = append(providers, static)
+	}
+	return providers, nil
+}
+
+// StaticProvider holds a list of target groups that never change.
+type StaticProvider struct {
+	targetGroups []*config.TargetGroup
+}
+
+// NewStaticProvider returns a StaticProvider configured with the given
+// target groups.
+func NewStaticProvider(groups []*config.TargetGroup) *StaticProvider {
+	for i, tg := range groups {
+		tg.Source = fmt.Sprintf("static:%d", i)
+	}
+	return &StaticProvider{groups}
+}
+
+// Run implements the TargetProvider interface.
+func (sd *StaticProvider) Run(ch chan<- *config.TargetGroup) {
+	for _, tg := range sd.targetGroups {
+		ch <- tg
+	}
+	close(ch) // This provider never sends any updates.
 }
 
-func (m *targetManager) Pools() map[string]*TargetPool {
-	m.Lock()
-	defer m.Unlock()
+// Stop implements the TargetProvider interface.
+func (sd *StaticProvider) Stop() {}
 
-	result := make(map[string]*TargetPool, len(m.poolsByJob))
-	for k, v := range m.poolsByJob {
-		result[k] = v
+// TargetGroups returns the provider's target groups.
+func (sd *StaticProvider) Sources() (srcs []string) {
+	for _, tg := range sd.targetGroups {
+		srcs = append(srcs, tg.Source)
 	}
-	return result
+	return
```

Comment:
I mean, it's short enough to allow naked returns (but still, avoid IMO), but in this case it's just a naked return at the end of the function. Can just be omitted, no?